File: engine/src/main/java/com/terralite/engine/physics/CollisionResponseSystem.java
package com.terralite.engine.physics;

import com.terralite.engine.entity.Entity;
import com.terralite.engine.simulation.SimulationTick;
import com.terralite.engine.simulation.WorldSimulationSystem;
import com.terralite.engine.world.World;

import java.util.Objects;

public final class CollisionResponseSystem implements WorldSimulationSystem {
    private final CollisionDetector detector;

    public CollisionResponseSystem() {
        this(new CollisionDetector());
    }

    public CollisionResponseSystem(CollisionDetector detector) {
        this.detector = Objects.requireNonNull(detector, "detector");
    }

    @Override
    public void tick(World world, SimulationTick tick) {
        for (CollisionPair collision : detector.detect(world)) {
            revertIfMoved(collision.first());
            revertIfMoved(collision.second());
        }
    }

    private static void revertIfMoved(Entity entity) {
        entity.get(PhysicsComponents.PREVIOUS_TRANSFORM).ifPresent(previous -> {
            entity.set(PhysicsComponents.TRANSFORM, previous);
            if (entity.has(PhysicsComponents.VELOCITY)) {
                entity.set(PhysicsComponents.VELOCITY, Velocity.ZERO);
            }
        });
    }
}

File: engine/src/commonMain/kotlin/com/terralite/engine/physics/CollisionResponseSystem.kt
package com.terralite.engine.physics

import com.terralite.engine.entity.Entity
import com.terralite.engine.simulation.SimulationTick
import com.terralite.engine.simulation.WorldSimulationSystem
import com.terralite.engine.world.World

class CollisionResponseSystem @JvmOverloads constructor(
    private val detector: CollisionDetector = CollisionDetector()
) : WorldSimulationSystem {
    override fun tick(world: World, tick: SimulationTick) {
        for (collision in detector.detect(world)) {
            revertIfMoved(collision.first)
            revertIfMoved(collision.second)
        }
    }

    private fun revertIfMoved(entity: Entity) {
        val previous = entity.get(PhysicsComponents.PREVIOUS_TRANSFORM) ?: return
        entity.set(PhysicsComponents.TRANSFORM, previous)
        if (entity.has(PhysicsComponents.VELOCITY)) {
            entity.set(PhysicsComponents.VELOCITY, Velocity.ZERO)
        }
    }
}
